File: iosApp/iosApp/DesignSystem/DsButton.swift
import SwiftUI
import Shared

/// DS Button — MUI-style API. Variants: text, contained, outlined.
/// Colors: primary, secondary, success, error, warning, info. Sizes: small, medium, large.
struct DsButton<Content: View>: View {
    let action: () -> Void
    let content: Content
    var variant: ButtonVariant
    var colorRole: ButtonColorRole
    var size: ButtonSize
    var fullWidth: Bool
    var disabled: Bool
    var loading: Bool

    init(
        action: @escaping () -> Void,
        variant: ButtonVariant = .contained,
        color: ButtonColorRole = .primary,
        size: ButtonSize = .medium,
        fullWidth: Bool = false,
        disabled: Bool = false,
        loading: Bool = false,
        @ViewBuilder content: () -> Content
    ) {
        self.action = action
        self.content = content()
        self.variant = variant
        self.colorRole = color
        self.size = size
        self.fullWidth = fullWidth
        self.disabled = disabled
        self.loading = loading
    }

    var body: some View {
        let tokens = ButtonTokens.shared
        let containerColor = Color(kmpArgb: tokens.getContainerColor(variant: variant, colorRole: colorRole))
        let contentColor = Color(kmpArgb: tokens.getContentColor(variant: variant, colorRole: colorRole))
        let borderColor = Color(kmpArgb: tokens.getBorderColor(variant: variant, colorRole: colorRole))
        let paddingH = tokens.getPaddingHorizontal(size: size)
        let paddingV = tokens.getPaddingVertical(size: size)
        let cornerRadius = tokens.getCornerRadius(size: size)

        let isOutlined = variant == .outlined
        let isText = variant == .text

        Button(action: action) {
            Group {
                if loading {
                    ProgressView()
                        .progressViewStyle(CircularProgressViewStyle(tint: contentColor))
                } else {
                    content
                }
            }
            .padding(.horizontal, paddingH)
            .padding(.vertical, paddingV)
            .frame(maxWidth: fullWidth ? .infinity : nil)
            .foregroundColor(contentColor)
            .background(isText || isOutlined ? Color.clear : containerColor)
            .overlay(
                isOutlined ? RoundedRectangle(cornerRadius: cornerRadius)
                    .stroke(borderColor, lineWidth: 1) : nil
            )
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
            .opacity(disabled || loading ? 0.5 : 1)
        }
        .buttonStyle(.plain)
        .disabled(disabled || loading)
    }
}

File: composeApp/src/androidMain/kotlin/com/appointy/designsystem/components/DsButton.kt
package com.appointy.designsystem.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import com.core.designsystem.tokens.components.button.ButtonColorRole
import com.core.designsystem.tokens.components.button.ButtonSize
import com.core.designsystem.tokens.components.button.ButtonVariant

/**
 * DS Button — MUI-style API with Slot pattern.
 * Variants: text, contained, outlined. Colors: primary, secondary, success, error, warning, info.
 * Sizes: small, medium, large. Optional startIcon/endIcon slots.
 */
@Composable
fun DsButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    variant: ButtonVariant = ButtonVariant.CONTAINED,
    color: ButtonColorRole = ButtonColorRole.PRIMARY,
    size: ButtonSize = ButtonSize.MEDIUM,
    fullWidth: Boolean = false,
    disableElevation: Boolean = false,
    loading: Boolean = false,
    startIcon: (@Composable () -> Unit)? = null,
    endIcon: (@Composable () -> Unit)? = null,
    content: @Composable () -> Unit
) {
    val shape = AppButtonDefaults.shape(size)
    val contentColor = AppButtonDefaults.contentColor(variant, color)
    val containerColor = AppButtonDefaults.containerColor(variant, color)
    val padding = AppButtonDefaults.contentPadding(size)
    val border = AppButtonDefaults.borderStroke(variant, color)
    val mod = if (fullWidth) modifier.fillMaxWidth() else modifier
    val elevation = if (disableElevation) ButtonDefaults.buttonElevation(0.dp, 0.dp, 0.dp) else ButtonDefaults.elevatedButtonElevation()

    @Composable
    fun ButtonSlot() {
        if (loading) {
            val loaderSize = when (size) {
                ButtonSize.SMALL -> 16.dp
                ButtonSize.MEDIUM -> 20.dp
                ButtonSize.LARGE -> 24.dp
            }
            CircularProgressIndicator(
                modifier = Modifier.size(loaderSize),
                color = contentColor,
                strokeWidth = 2.dp
            )
        } else {
            Row(
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.Center
            ) {
                startIcon?.invoke()
                content()
                endIcon?.invoke()
            }
        }
    }

    val isEnabled = enabled && !loading

    when (variant) {
        ButtonVariant.TEXT -> TextButton(
            onClick = onClick,
            modifier = mod,
            enabled = isEnabled,
            shape = shape,
            colors = ButtonDefaults.textButtonColors(contentColor = contentColor),
            contentPadding = padding
        ) { ButtonSlot() }
        ButtonVariant.OUTLINED -> OutlinedButton(
            onClick = onClick,
            modifier = mod,
            enabled = isEnabled,
            shape = shape,
            colors = ButtonDefaults.outlinedButtonColors(contentColor = contentColor),
            border = border!!,
            contentPadding = padding
        ) { ButtonSlot() }
        ButtonVariant.CONTAINED -> Button(
            onClick = onClick,
            modifier = mod,
            enabled = isEnabled,
            shape = shape,
            colors = ButtonDefaults.buttonColors(
                containerColor = containerColor,
                contentColor = contentColor
            ),
            elevation = elevation,
            contentPadding = padding
        ) { ButtonSlot() }
    }
}
